### ISSUE TITLE
clang-14 crash compiling templates (works fine with g++/visual studio).

### ISSUE THREAD

**Issue by LewisPringle** [llvm/llvm-project#62785]
[VariantValue-587e15.zip](https://github.com/llvm/llvm-project/files/11501764/VariantValue-587e15.zip)

**Comment by LewisPringle** [comment]
using 'typename' instead of Common::IEqualsComparer<T> in template spec makes the problem go away.

**Comment by EugeneZelenko** [comment]
Could you please try 16 or `main` branch? https://godbolt.org should be helpful.

**Comment by DimitryAndric** [comment]
I think this is another case of babdef27c503c0bbbcc017e9f88affddda90ea4e ("Re-apply "Deferred Concept Instantiation Implementation") by @erichkeane. This only landed in clang 16, and I think there were also follow-ups.

In any case, this should be fixed in 16 (16.0.0, even?) and main.

**Comment by DimitryAndric** [comment]
Reduced:

```c++
// clang -cc1 -triple x86_64-- -S -std=c++20 VariantValue-min.cpp
namespace Common {
enum ComparisonRelationType { eEquals };
template <typename> struct ExtractComparisonTraits {
  static constexpr ComparisonRelationType kComparisonRelationKind = eEquals;
};
template <typename, typename> constexpr bool IsEqualsComparer();
template <typename COMPARER, typename ARG_T>
concept IEqualsComparer = IsEqualsComparer<COMPARER, ARG_T>();
template <typename COMPARER, typename> constexpr bool IsEqualsComparer() {
  return ExtractComparisonTraits<COMPARER>::kComparisonRelationKind == eEquals;
}
} // namespace Common
namespace Traversal {
template <typename T> class Iterable {
public:
  template <Common::IEqualsComparer<T>> struct SequentialEqualsComparer;
};
template <typename T>
template <Common::IEqualsComparer<T> T_EQUALS_COMPARER>
struct Iterable<T>::SequentialEqualsComparer {
  constexpr SequentialEqualsComparer(const T_EQUALS_COMPARER & = {},
                                     bool = false);
};
template <typename T>
template <Common::IEqualsComparer<T> T_EQUALS_COMPARER>
constexpr Iterable<T>::SequentialEqualsComparer<
    T_EQUALS_COMPARER>::SequentialEqualsComparer(const T_EQUALS_COMPARER &,
                                                 bool) {}
} // namespace Traversal
namespace Containers {
using Traversal::Iterable;
template <typename T> class Sequence {
  using inherited = Iterable<T>;
  using value_type = inherited;
  template <typename T_EQUALS_COMPARER = T>
  using EqualsComparer =
      typename value_type ::template SequentialEqualsComparer<
          T_EQUALS_COMPARER>;

public:
  bool operator==(const Sequence &) const;
};
template <typename T> bool Sequence<T>::operator==(const Sequence &) const {
  EqualsComparer<>{};
}
} // namespace Containers
namespace Stroika::Foundation::DataExchange {
using Containers::Sequence;
class VariantValue {
  struct EqualsComparer;
};
struct VariantValue::EqualsComparer {
  bool operator()(const VariantValue &, const VariantValue &) const;
};
} // namespace Stroika::Foundation::DataExchange
bool Stroika::Foundation::DataExchange::VariantValue::EqualsComparer::
operator()(const VariantValue &, const VariantValue &) const {
  Sequence<VariantValue>() == Sequence<VariantValue>();
}
```

**Comment by DimitryAndric** [comment]
Since there's a workaround, and it's been fixed in 16.0, let's close this, as I don't think it is very likely that any of the applicable fixes will be backported to 15.0 or 14.0.
